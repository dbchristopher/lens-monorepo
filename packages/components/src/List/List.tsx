import React, { FunctionComponent, Ref } from 'react';
import styled, { css, StyledComponent } from 'styled-components';
import { BoxProps } from '../Box/Box';
import Box from '../Box';

export type ListTypes = 'bullet' | 'number' | 'letter';

export interface ListProps
  extends BoxProps<HTMLOListElement | HTMLUListElement> {
  type?: ListTypes;
  nomarker?: boolean;
}

type ComponentType = FunctionComponent<ListProps>;
type StyledComponentType = StyledComponent<ComponentType, ListProps>;

/**
 * List are stacked groups of related content that can be useful in many contexts.
 */
const InternalList: ComponentType = ({ type, ...props }) => {
  const pl = props.nomarker ? 'none' : 'medium';
  delete props.nomarker;
  switch (type) {
    case 'number':
    case 'letter':
      return (
        <Box is="ol" m="none" pl={pl} {...props}>
          {props.children}
        </Box>
      );
    case 'bullet':
      return (
        <Box is="ul" m="none" pl={pl} {...props}>
          {props.children}
        </Box>
      );
    default:
      return (
        <Box is="ul" m="none" p="none" {...props}>
          {props.children}
        </Box>
      );
  }
};

function listStyleType(props: ListProps) {
  switch (props.nomarker ? undefined : props.type) {
    case 'bullet':
      return css`
        list-style-type: disc;
      `;
    case 'number':
      return css`
        list-style-type: decimal;
      `;
    case 'letter':
      return css`
        list-style-type: upper-alpha;
      `;
    default:
      return css`
        list-style-type: none;
      `;
  }
}

const ListFactory = React.forwardRef<StyledComponentType, ListProps>(
  (props: ListProps, ref: Ref<StyledComponentType>) => (
    <InternalList ref={ref} {...props} />
  )
);

const List = styled<ComponentType>(ListFactory)`
  ${listStyleType};
`;

export default List;
