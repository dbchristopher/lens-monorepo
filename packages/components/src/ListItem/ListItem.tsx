import React, { FunctionComponent, Ref } from 'react'
import styled, { StyledComponent } from 'styled-components'
import { BoxProps } from '../Box/Box'
import Box from '../Box'

export type ListItemProps = BoxProps<HTMLLIElement>
type ComponentType = FunctionComponent<ListItemProps>
type StyledComponentType = StyledComponent<ComponentType, ListItemProps>

const InternalListItem: ComponentType = props => (
  <Box is="li" mb="xxsmall" {...props}>
    {props.children}
  </Box>
)

const ListItemFactory = React.forwardRef<StyledComponentType, ListItemProps>(
  (props: ListItemProps, ref: Ref<StyledComponentType>) => (
    <InternalListItem ref={ref} {...props} />
  )
)

const ListItem = styled<ComponentType>(ListItemFactory)``

export default ListItem
