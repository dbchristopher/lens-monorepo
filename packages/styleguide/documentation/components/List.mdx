---
name: List
menu: Components
---

import { List, ListItem, Link, Code } from 'looker-lens-components';
import { Playground } from 'docz';

# The Default List

By default a `<List />` component will render as an unordered list, `<ul>` tag, with no marker. Each item in the list should be wrapped in the `<ListItem />` component. Below is an example of creating a list of links using the default list behavior.

<Playground>
  <List>
    <ListItem>
      <Link href="/#!/Card">Lens Card</Link>
    </ListItem>
    <ListItem>
      <Link href="/#!/Link">Lens Link</Link>
    </ListItem>
    <ListItem>
      <Link href="/#!/Heading">Lens Heading</Link>
    </ListItem>
  </List>
</Playground>

## List Types

Often times you want to have a marker before each item in your list, you can use the `type` property to change list marker and also the semantic html tag used to wrap the list.

<Playground>
  <List>
    <ListItem>An unordered list</ListItem>
    <ListItem>
      Wraps in a <Code>&lt;ul&gt;</Code> tag
    </ListItem>
    <ListItem>No marker</ListItem>
  </List>

<List type="bullet">
  <ListItem>An unordered list</ListItem>
  <ListItem>
    Wraps in a <Code>&lt;ul&gt;</Code> tag
  </ListItem>
  <ListItem>Bullet marker</ListItem>
</List>

<List type="number">
  <ListItem>An orderd list</ListItem>
  <ListItem>
    Wraps in a <Code element="code">&lt;ol&gt;</Code> tag
  </ListItem>
  <ListItem>Number marker</ListItem>
</List>

  <List type="letter">
    <ListItem>An ordered list</ListItem>
    <ListItem>
      Wraps in a <Code>&lt;ol&gt;</Code> tag
    </ListItem>
    <ListItem>Alpha marker</ListItem>
  </List>
</Playground>

## Hiding the List marker

If for semantic reasons you need an ordered list but do not want to display the marker, you can control that with the `nomarker` property.

<Playground>
  <List type="number" nomarker>
    <ListItem>I am an ordered list</ListItem>
    <ListItem>But I do not have a marker</ListItem>
    <ListItem>That is neat</ListItem>
  </List>
</Playground>
