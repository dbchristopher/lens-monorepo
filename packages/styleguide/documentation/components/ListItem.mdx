---
name: List Item
menu: Components
---

# List Item

import { List, ListItem } from 'looker-lens-components';
import { Playground } from 'docz';

This component renders an `<li>` tag inside of a list

<Playground>
  <List>
    <ListItem>Item 1</ListItem>
    <ListItem>Item 2</ListItem>
    <ListItem>Item 3</ListItem>
  </List>
</Playground>
