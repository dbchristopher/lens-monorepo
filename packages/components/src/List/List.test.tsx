import React from 'react';
import { createWithTheme } from '../../../test/utils/create_with_theme';
import { List } from './List';
import { ListItem } from '../ListItem/ListItem';

import 'jest-styled-components';

test('A default List, should be a ul', () => {
  const component = createWithTheme(
    <List>
      <ListItem>🥑</ListItem>
      <ListItem>🍕</ListItem>
      <ListItem>🥨</ListItem>
    </List>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A bulleted List', () => {
  const component = createWithTheme(
    <List type="bullet">
      <ListItem>🥑</ListItem>
      <ListItem>🍕</ListItem>
      <ListItem>🥨</ListItem>
    </List>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A numerically ordered List', () => {
  const component = createWithTheme(
    <List type="number">
      <ListItem>🥑</ListItem>
      <ListItem>🍕</ListItem>
      <ListItem>🥨</ListItem>
    </List>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A numerically ordered List marked as nomarker', () => {
  const component = createWithTheme(
    <List type="number" nomarker>
      <ListItem>🥑</ListItem>
      <ListItem>🍕</ListItem>
      <ListItem>🥨</ListItem>
    </List>
  );
  const tree = component.toJSON();
  expect(tree).toHaveStyleRule('list-style-type', 'none');
  expect(tree).toHaveStyleRule('padding-left', '0rem');
  expect(tree).toMatchSnapshot();
});
