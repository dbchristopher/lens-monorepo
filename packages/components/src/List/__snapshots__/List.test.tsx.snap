// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A bulleted List 1`] = `
.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  margin-bottom: 0.25rem;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  margin: 0rem;
  padding-left: 1rem;
}

.c0 {
  list-style-type: disc;
}

<ul
  className="c0 c1"
>
  <li
    className="c2"
  >
    ü•ë
  </li>
  <li
    className="c2"
  >
    üçï
  </li>
  <li
    className="c2"
  >
    ü•®
  </li>
</ul>
`;

exports[`A default List, should be a ul 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  margin: 0rem;
  padding: 0rem;
}

.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  margin-bottom: 0.25rem;
}

.c0 {
  list-style-type: none;
}

<ul
  className="c0 c1"
>
  <li
    className="c2"
  >
    ü•ë
  </li>
  <li
    className="c2"
  >
    üçï
  </li>
  <li
    className="c2"
  >
    ü•®
  </li>
</ul>
`;

exports[`A numerically ordered List 1`] = `
.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  margin-bottom: 0.25rem;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  margin: 0rem;
  padding-left: 1rem;
}

.c0 {
  list-style-type: decimal;
}

<ol
  className="c0 c1"
>
  <li
    className="c2"
  >
    ü•ë
  </li>
  <li
    className="c2"
  >
    üçï
  </li>
  <li
    className="c2"
  >
    ü•®
  </li>
</ol>
`;

exports[`A numerically ordered List marked as nomarker 1`] = `
.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  margin-bottom: 0.25rem;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  margin: 0rem;
  padding-left: 0rem;
}

.c0 {
  list-style-type: none;
}

<ol
  className="c0 c1"
>
  <li
    className="c2"
  >
    ü•ë
  </li>
  <li
    className="c2"
  >
    üçï
  </li>
  <li
    className="c2"
  >
    ü•®
  </li>
</ol>
`;
